bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,595 @@
+    private static final String HUB_REST_SET_BRIGHTNESS = "{\"ID\":%d,\"Service\":\"SetZoneProperties\",\"ZID\":%d,\"PropertyList\":{\"PowerLevel\":%d}}\0";
+    private static final String HUB_REST_REQUEST_STATE = "{\"ID\":%d,\"Service\":\"ReportZoneProperties\",\"ZID\":%d}\0";
+    private static final String HUB_REST_REQUEST_ZONES = "{\"ID\":%d,\"Service\":\"ListZones\"}\0";
+    private static final String HUB_REST_REQUEST_MACADDRESS = "{\"ID\":%d,\"Service\":\"SystemInfo\"}\0";
+    private static final String HUB_TOKEN_SERVICE = "Service";
+    private static final String HUB_TOKEN_ZID = "ZID";
+    private static final String HUB_TOKEN_PROPERTY_LIST = "PropertyList";
+    private static final String HUB_TOKEN_DEVICE_TYPE = "DeviceType";
+    private static final String HUB_TOKEN_SWITCH = "Switch";
+    private static final String HUB_TOKEN_DIMMER = "Dimmer";
+    private static final String HUB_TOKEN_NAME = "Name";
+    private static final String HUB_TOKEN_POWER = "Power";
+    private static final String HUB_TOKEN_POWER_LEVEL = "PowerLevel";
+    private static final String HUB_TOKEN_MAC_ADDRESS = "MACAddress";
+    private static final String HUB_TOKEN_ZONE_LIST = "ZoneList";
+    private static final String HUB_SERVICE_REPORT_ZONE_PROPERTIES = "ReportZoneProperties";
+    private static final String HUB_SERVICE_ZONE_PROPERTIES_CHANGED = "ZonePropertiesChanged";
+    private static final String HUB_SERVICE_LIST_ZONE = "ListZones";
+    private static final String HUB_SERVICE_SYSTEM_INFO = "SystemInfo";
+
+    private @Nullable Future<?> hubController;
+    private String hubHost;
+    private int hubPort;
+    private @Nullable Socket hubSocket;
+    private @Nullable PrintStream hubOut;
+    private @Nullable InputStreamReader hubInReader;
+    private @Nullable JsonStreamParser hubIn;
+    private CompletableFuture<@Nullable Void> hubControllerConnected;
+    private int hubReconnectSleep; // Sleep time before we attempt re-connect
+    private ScheduledExecutorService scheduler;
+
+    private volatile boolean stop; // Stop the controller as soon as possible
+    private volatile boolean stopWhenCommandsServed; // Stop the controller once all pending commands have been served
+    private volatile long stopTimestamp = 0; // Stop the controller after this timestamp
+
+    // When we submit commmands to the hub we don't correlate commands and responses. We simply use the first available
+    // response that answers our question. For that we maintain lists of all pending commands.
+    private Map<Integer, CompletableFuture<AdorneDeviceState>> stateCommands;
+    private @Nullable CompletableFuture<List<Integer>> zoneCommand;
+    private @Nullable CompletableFuture<String> macAddressCommand;
+    private int commandId; // We assign increasing command ids to all rest commands to the hub for easier
```

Comment:
Make this an AtomicInteger and just make calls to incrementAndGet() every time you want a new id. This approach is a much more thread safe.

Reply:
I hadn't considered parallel commands, but you are correct - I need to. However, I can't use incrementAndGet as that overflows into negative values. So I used updateAndGet. Looking at the result wouldn't a simple lock be much easier to read though?

Reply:
I think using an AtomicInteger displays clearer intentions as to how the field is used and accessed vs using basic synchronization. Also your use case happens to be exactly what AtomicInteger was designed for. I have no problem with you using updateAndGet to keep values positive. 👍 
But just to keep things in perspective, if you incremented the id once every second it would take over 63 years before it overflowed. 😮

Reply:
OpenHAB will be around then, right? :-)